plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java

```diff
@@ -41,43 +42,75 @@ public PipelineExecutionService(PipelineStore store, ThreadPool threadPool) {
-            String index = indexRequest.index();
-            String type = indexRequest.type();
-            String id = indexRequest.id();
-            String routing = indexRequest.routing();
-            String parent = indexRequest.parent();
-            String timestamp = indexRequest.timestamp();
-            String ttl = indexRequest.ttl() == null ? null : indexRequest.ttl().toString();
-            Map<String, Object> sourceAsMap = indexRequest.sourceAsMap();
-            IngestDocument ingestDocument = new IngestDocument(index, type, id, routing, parent, timestamp, ttl, sourceAsMap);
             try {
-                pipeline.execute(ingestDocument);
-                Map<IngestDocument.MetaData, String> metadataMap = ingestDocument.extractMetadata();
-                //it's fine to set all metadata fields all the time, as ingest document holds their starting values
-                //before ingestion, which might also get modified during ingestion.
-                indexRequest.index(metadataMap.get(IngestDocument.MetaData.INDEX));
-                indexRequest.type(metadataMap.get(IngestDocument.MetaData.TYPE));
-                indexRequest.id(metadataMap.get(IngestDocument.MetaData.ID));
-                indexRequest.routing(metadataMap.get(IngestDocument.MetaData.ROUTING));
-                indexRequest.parent(metadataMap.get(IngestDocument.MetaData.PARENT));
-                indexRequest.timestamp(metadataMap.get(IngestDocument.MetaData.TIMESTAMP));
-                indexRequest.ttl(metadataMap.get(IngestDocument.MetaData.TTL));
-                indexRequest.source(ingestDocument.getSourceAndMetadata());
-                listener.onResponse(null);
-            } catch (Throwable e) {
-                listener.onFailure(e);
+                innerExecute(request, pipeline);
+                completionHandler.accept(true);
+            } catch (Exception e) {
+                failureHandler.accept(e);
+            }
+        });
+    }
+
+    public void execute(Iterable<ActionRequest> indexRequests, String pipelineId,
+                        Consumer<Throwable> itemFailureHandler, Consumer<Boolean> completionHandler) {
+        Pipeline pipeline = getPipeline(pipelineId);
+        threadPool.executor(THREAD_POOL_NAME).execute(() -> {
+            Throwable lastThrowable = null;
+            for (ActionRequest actionRequest : indexRequests) {
+                if ((actionRequest instanceof IndexRequest) == false) {
+                    continue;
+                }
+
+                IndexRequest indexRequest = (IndexRequest) actionRequest;
+                try {
+                    innerExecute(indexRequest, pipeline);
+                } catch (Throwable e) {
+                    lastThrowable = e;
+                    if (itemFailureHandler != null) {
+                        itemFailureHandler.accept(e);
+                    }
+                }
             }
+            completionHandler.accept(lastThrowable == null);
```

Comment:
ok I would always pass down true then otherwise it feels like we should do something with it. Consumer<Void> would be better in that sense but then you need to provide null which is even worse to see...